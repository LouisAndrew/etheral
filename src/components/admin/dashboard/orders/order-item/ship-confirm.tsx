import React, { useState } from 'react';
import { useForm } from 'react-hook-form';

import { Box, Button } from 'rebass';
import { Input, Label } from '@rebass/forms';

type Props = {
    confirmShipping: (data: Inputs) => void;
};

export type Inputs = {
    shippingDate: Date;
    trackingNum: string;
    shippedBy: string;
};

const ShippingConfirmation: React.FC<Props> = ({ confirmShipping }) => {
    const [viewForm, setViewForm] = useState(false);
    const { register, handleSubmit } = useForm();

    const submit = (data: Inputs) => {
        confirmShipping(data);
        setViewForm(false);
    };

    return viewForm ? (
        <Box
            as="form"
            data-testid="shipping-info-form"
            css={`
                & > input,
                & > label {
                    font-family: 'Montserrat', sans-serif;
                }
            `}
            onSubmit={handleSubmit(submit)}
        >
            <Label mb={[2, 2, 3]} htmlFor="shipping-date">
                Shipping date
            </Label>
            <Input
                mb={[4, 4, 5]}
                id="shipping-date"
                name="shippingDate"
                type="date"
                ref={register}
            />
            <Label mb={[2, 2, 3]} htmlFor="tracking-num">
                Tracking number
            </Label>
            <Input
                mb={[4, 4, 5]}
                id="tracking-num"
                name="trackingNum"
                ref={register}
            />
            <Label mb={[2, 2, 3]} htmlFor="shipped-by">
                Shipped by
            </Label>
            <Input
                mb={[4, 4, 5]}
                id="shipped-by"
                name="shippedBy"
                ref={register}
            />
            <Input
                value="Confirm"
                type="submit"
                bg="brown.2"
                color="#fff"
                css={`
                    outline: none !important;
                    &:hover {
                        cursor: pointer;
                    }
                `}
                width="fit-content"
            />
        </Box>
    ) : (
        <Button variant="primary" onClick={() => setViewForm(true)}>
            Confirm Shipping
        </Button>
    );
};

export default ShippingConfirmation;
