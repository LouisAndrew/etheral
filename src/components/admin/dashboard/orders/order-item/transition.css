.info-enter {
    opacity: 0;
    max-height: 0;
}
.info-enter-active {
    opacity: 1;
    max-height: 0;
}
.info-exit {
    opacity: 1;
}
.info-exit-active {
    opacity: 0;
}
