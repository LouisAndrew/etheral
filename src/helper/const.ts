import { Product } from './schema/product';
import { Order } from './schema/order';

const mockProduct1: Product = {
    pid: '1234',
    name: 'Mock product',
    slug: 'ctg1/product1',
    category: 'ctg1',
    idrPrice: 23000,
    collection: 'fall',
    productImages: [],
    urls: [],
    weight: 20,
    amount: 12,
};

export const mockProducts: Product[] = [mockProduct1];

export const mockOrder: Order = {
    oid: 'asdadwda',
    buyerName: 'Jane Doe',
    buyerEmail: 'jane@doe',
    buyerPhone: 123412412,
    buyerUId: 'admin',
    buyerAddr: 'Storkowerstrasse',
    buyerPostal: 1234143,
    total: 32000,
    currency: 'IDR',
    date: new Date(),
    via: 'va',
    channel: 'bni',
    products: [{ pid: mockProduct1.pid, amount: 12 }],
    paid: false,
    delivered: false,
    transactionData: {
        sessionId: '1234',
        paymentNo: '12',
        paymentName: 'Jane Doe - BNI',
        expired: '1231241412',
        fee: 3000,
    },
    shippingMethod: 'OKE',
};

export const mockOrders: Order[] = [mockOrder];

export const mockOrderShipped: Order = {
    ...mockOrder,
    delivered: true,
    shippingData: {
        shippedDate: new Date(),
        trackingCode: '12342',
        shippedBy: 'admin1',
    },
};
