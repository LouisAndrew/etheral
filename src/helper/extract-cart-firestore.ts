import { intersectionBy } from 'lodash';

import { SetCartArgs } from 'state/actions/cart';
import { InCart } from './schema/firebase-user';
import { Product } from './schema/product';

export default ({
    firestoreCartData,
    allProducts,
}: {
    firestoreCartData: InCart;
    allProducts: Product[];
}): SetCartArgs => {
    const inCartProduct = intersectionBy(allProducts, firestoreCartData, 'pid');
    return inCartProduct.map(product => ({
        product,
        amount: intersectionBy(firestoreCartData, [product], 'pid')[0].amount, // get the intersection between the product -> pid is unique, so the function is guaranteed to return one item.
    }));
};
