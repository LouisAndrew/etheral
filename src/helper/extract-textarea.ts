import { snakeCase } from 'lodash';

/**
 * function to extract options from a textarea
 * @param textarea textarea input (e.g sizes: S, M, L)
 */
export const extractTextArea = (textarea: string) =>
    textarea.split(',').map(text => snakeCase(text).toUpperCase());
