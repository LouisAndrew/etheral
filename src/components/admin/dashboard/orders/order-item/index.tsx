import React, { useEffect, useState } from 'react';
import { OrderItem as OrderItemEl } from './order-item';
import firebase from 'gatsby-plugin-firebase';

import { Order } from 'helper/schema/order';
import useAllProducts from 'helper/use-all-products';
import { Inputs } from './ship-confirm';

type Props = {
    order: Order;
};

const OrderItem: React.FC<Props> = ({ order }) => {
    const [db, setDb] = useState<firebase.firestore.Firestore | undefined>(
        undefined
    );

    const allProducts = useAllProducts();

    useEffect(() => {
        setDb(firebase.firestore());
    }, []);

    const updateShipping = (data: Inputs) => {
        if (db) {
            db.collection('order')
                .doc(order.oid)
                .update({
                    shippingData: {
                        ...data,
                        shippingDate: new Date(data.shippingDate),
                    },
                    delivered: true,
                });
        }
    };

    return db ? (
        <OrderItemEl
            order={order}
            allProducts={allProducts}
            updateShipping={updateShipping}
        />
    ) : (
        <></>
    );
};

export default OrderItem;
