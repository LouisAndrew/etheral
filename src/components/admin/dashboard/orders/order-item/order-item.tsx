import React from 'react';
import { findIndex, get } from 'lodash';
import Img, { FixedObject } from 'gatsby-image';

import { Box, Flex, Heading, Text } from 'rebass';
import { InlineIcon } from '@iconify/react';
import priceTag3Fill from '@iconify/icons-ri/price-tag-3-fill';

import { getDate } from 'helper/get-date';
import { Order } from 'helper/schema/order';
import { Product } from 'helper/schema/product';
import useAllProductImages from 'helper/use-all-product-images';
import StatusBadge, { BadgeTypes } from '../status-badge';
import { InCart } from 'helper/schema/firebase-user';
import ShippingConfirmation, { Inputs } from './ship-confirm';

import { theme } from 'styles';

import './transition.css';

type Props = {
    order: Order;
    allProducts: Product[];
    updateShipping: (data: Inputs) => void;
};

const OrderItem: React.FC<Props> = ({ order, allProducts, updateShipping }) => {
    const { extractImgs } = useAllProductImages();

    const spanStyle = {
        marginLeft: '12px',
        fontWeight: 600,
        color: theme.colors.black[0],
    };

    // render each product within an order.
    const renderProduct = (orderedProduct: InCart) => {
        const index = findIndex(allProducts, o => o.pid === orderedProduct.pid);
        const product = allProducts[index];

        const { xs }: any = extractImgs(product, false);
        return index !== -1 ? (
            <Flex
                alignItems="center"
                py={[1]}
                sx={{
                    borderColor: 'black.1',
                    borderStyle: 'solid',
                    borderWidth: 0,
                    borderBottomWidth: 1,
                }}
            >
                <Img
                    fixed={
                        get(xs, 'img[0].childImageSharp.fixed') as FixedObject
                    }
                />
                <Text variant="productPrice" ml={[5]} my={[2]}>
                    {orderedProduct.amount} x {product.name}
                    {product.discountPercentage > 0 && (
                        <Box as="span" display="inline-block">
                            <Text
                                color="misc.discount"
                                fontWeight="bold"
                                ml={[4]}
                                sx={{ '& svg': { mr: [2] } }}
                            >
                                <InlineIcon
                                    icon={priceTag3Fill}
                                    color={theme.colors.misc.discount}
                                />
                                {product.discountPercentage}%
                            </Text>
                        </Box>
                    )}
                </Text>
            </Flex>
        ) : (
            <></>
        );
    };

    // updates UI state and order data on db.
    const confirmShipping = (data: Inputs) => {
        updateShipping(data);
    };

    const {
        products,
        paid,
        delivered,
        date,
        buyerName,
        buyerPhone,
        buyerEmail,
        buyerUId,
        via,
        channel,
        currency,
        total,
        buyerAddr,
        buyerPostal,
        shippingMethod,
        shippingData,
        discount,
        discountCode,
    } = order;

    return (
        <Box mt={[5]} mx={[3]}>
            <Heading fontFamily="body" fontWeight="body" color="black.0">
                Order ID: <strong>{order.oid}</strong>
            </Heading>

            {/* status badges */}
            <Flex className="status">
                <StatusBadge type={BadgeTypes.PAYMENT} paid={paid} />
                <StatusBadge type={BadgeTypes.SHIPPING} shipped={delivered} />
                <StatusBadge type={BadgeTypes.DATE} date={date} />
            </Flex>

            {/* products */}
            <Box
                bg="white.2"
                py={[4]}
                pr={[3]}
                pl={[4, 4, 5]}
                my={[5]}
                sx={{
                    borderColor: 'black.1',
                    borderStyle: 'solid',
                    borderWidth: 1,
                }}
            >
                <Heading as="h4" variant="adminOrderBody">
                    Products
                </Heading>
                <Box>
                    {products.map(product => (
                        <React.Fragment key={product.pid}>
                            {renderProduct(product)}
                        </React.Fragment>
                    ))}
                </Box>

                {/* render discount infos. */}
                {discountCode && (
                    <Box my={[4]}>
                        <Heading as="h4" variant="adminOrderBody">
                            Discount
                        </Heading>
                        <Box mt={[3]}>
                            <Text variant="productPrice">
                                Code:{' '}
                                <span style={{ fontWeight: 'bold' }}>
                                    {discountCode}
                                </span>
                            </Text>
                            <Text variant="productPrice">
                                Discount:{' '}
                                <span
                                    style={{
                                        fontWeight: 'bold',
                                        color: theme.colors.misc.discount,
                                    }}
                                >
                                    {discount}%
                                </span>
                            </Text>
                        </Box>
                    </Box>
                )}

                <Text variant="bodyMedium" mt={[5]} mr={[3]} textAlign="right">
                    Payment via{' '}
                    <span style={spanStyle}>
                        {via} - {channel}
                    </span>
                </Text>
                <Text variant="bodyMedium" mr={[3]} textAlign="right">
                    Subtotal:
                    <span style={spanStyle}>
                        {currency}
                        {'  '}
                        {total}
                    </span>
                </Text>
            </Box>

            {/* Customer info */}
            <Box overflow="hidden" my={[5]}>
                <Heading as="h4" variant="adminOrderBody">
                    Buyer&apos;s info
                </Heading>
                <Box ml={[3, 3, 4]}>
                    <Text variant="bodyMedium">
                        Name: <span style={spanStyle}>{buyerName}</span>
                    </Text>
                    <Text variant="bodyMedium">
                        Email: <span style={spanStyle}>{buyerEmail}</span>
                    </Text>
                    <Text variant="bodyMedium">
                        Phone: <span style={spanStyle}>{buyerPhone}</span>
                    </Text>
                    <Text variant="bodyMedium">
                        Buyer UserID:{' '}
                        <span style={spanStyle}>
                            {buyerUId === 'admin' ? 'None' : buyerUId}
                        </span>
                    </Text>
                </Box>
            </Box>

            {/* ship-to infos */}
            <Box my={[5]}>
                <Heading as="h4" variant="adminOrderBody">
                    Shipping info
                </Heading>
                <Box ml={[3, 3, 4]}>
                    <Text variant="bodyMedium">
                        Address: <span style={spanStyle}>{buyerAddr}</span>
                    </Text>
                    <Text variant="bodyMedium">
                        Postal code:{' '}
                        <span style={spanStyle}>{buyerPostal}</span>
                    </Text>
                    <Text variant="bodyMedium">
                        Shipping method:{' '}
                        <span style={spanStyle}>{shippingMethod}</span>
                    </Text>
                </Box>
            </Box>

            {/* confirm shipping. */}
            {delivered && shippingData ? (
                <>
                    <Heading as="h4" variant="adminOrderBody">
                        Shiping Informations
                    </Heading>
                    <Box data-testid="shipping-info" ml={[3, 3, 4]} mb={[5]}>
                        {shippingData.shippingDate && (
                            <Text variant="bodyMedium">
                                Shipped date:{' '}
                                <span style={spanStyle}>
                                    {getDate(shippingData.shippingDate)}
                                </span>
                            </Text>
                        )}
                        <Text variant="bodyMedium">
                            Tracking number:{' '}
                            <span style={spanStyle}>
                                {shippingData.trackingNum}
                            </span>
                        </Text>
                        <Text variant="bodyMedium">
                            Shipped by:{' '}
                            <span style={spanStyle}>
                                {shippingData.shippedBy}
                            </span>
                        </Text>
                    </Box>
                </>
            ) : (
                <ShippingConfirmation confirmShipping={confirmShipping} />
            )}
        </Box>
    );
};

export { OrderItem };
