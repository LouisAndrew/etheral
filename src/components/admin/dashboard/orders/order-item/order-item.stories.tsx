import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { OrderItem } from './order-item';
import { mockOrder, mockProducts, mockOrderShipped } from 'helper/const';
// import more addons

const story = storiesOf('Components.Admin.Dashboard.Orders.Order Item', module);

story
    .add('Not shipped order', () => (
        <OrderItem
            order={mockOrder}
            allProducts={mockProducts}
            updateShipping={action('update-shipping')}
        />
    ))
    .add('Shipped order', () => (
        <OrderItem
            order={mockOrderShipped}
            allProducts={mockProducts}
            updateShipping={action('update-shipping')}
        />
    ));
