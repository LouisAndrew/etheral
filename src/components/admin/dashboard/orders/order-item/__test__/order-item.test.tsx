import React from 'react';
import ReactDOM from 'react-dom';
// import renderer from 'react-test-renderer';

// import { render, cleanup } from '@testing-library/react';
import { cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
// import userEvent from '@testing-library/user-event';

import { OrderItem } from '../order-item';
// import { mockOrder, mockOrderShipped, mockProducts } from 'helper/const';
import { mockOrder, mockProducts } from 'helper/const';

// import * as hook from 'helper/use-all-product-images';

// const useAllProductImages = jest.spyOn(hook, 'default');

describe('OrderItem Component', () => {
    const mockUpdateShipping = jest.fn(() => {});
    const mockHideOrder = jest.fn(() => {});

    const el: React.ReactElement = (
        <OrderItem
            order={mockOrder}
            allProducts={mockProducts}
            updateShipping={mockUpdateShipping}
            hideOrder={mockHideOrder}
        />
    );

    // can't test behavior now, as I don't know how to mock custom hook properly

    // const elShipped: React.ReactElement = (
    //     <OrderItem
    //         order={mockOrderShipped}
    //         allProducts={mockProducts}
    //         updateShipping={mockUpdateShipping}
    //     />
    // );

    afterEach(cleanup);
    // beforeEach(() => {
    //     const emptyData = {
    //         img: [
    //             {
    //                 childImageSharp: {
    //                     fixed: [],
    //                     fluid: [],
    //                 },
    //                 url: '',
    //             },
    //         ],
    //         pid: '',
    //     };

    //     const sizesWithEmptyData = {
    //         xs: [emptyData],
    //         s: [emptyData],
    //         m: [emptyData],
    //         l: [emptyData],
    //         xl: [emptyData],
    //     };

    //     useAllProductImages.mockImplementationOnce(() => ({
    //         extractImgs: () => ({
    //             xs: emptyData,
    //             s: emptyData,
    //             m: emptyData,
    //             l: emptyData,
    //             xl: emptyData,
    //         }),
    //         ...sizesWithEmptyData,
    //     }));
    // });

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(el, div);
    });

    // behaviorial tests => provide set timeout to give element some time to render.

    // it('should display a form asking for shipping info whenever admin checked the delivered checkbox', () => {
    //     const { queryByText, queryByTestId, getByRole } = render(el);
    //     const checkShippingInfoForm = () => {
    //         const ShippingDataForm = queryByTestId('shipping-info-form');
    //         if (ShippingDataForm) {
    //             expect(ShippingDataForm).toBeInTheDocument();
    //         } else {
    //             fail();
    //         }
    //     };

    //     const checkForShippingConfirmation = () => {
    //         const NotShippedKeyword = queryByText(
    //             'Status: Not shipped'
    //         ) as HTMLInputElement;

    //         if (NotShippedKeyword) {
    //             expect(NotShippedKeyword).toBeInTheDocument();
    //         } else {
    //             fail();
    //         }

    //         const ConfirmShipButton = getByRole('button');

    //         userEvent.click(ConfirmShipButton);

    //         setTimeout(checkShippingInfoForm, 300);
    //     };

    //     setTimeout(checkForShippingConfirmation, 300);
    // });

    // it('should not display the shipping infos if the order has already shipped', () => {
    //     const { queryByTestId } = render(elShipped);
    //     setTimeout(() => {
    //         const ShippingInfoCard = queryByTestId('shipping-info');

    //         if (!ShippingInfoCard) {
    //             fail();
    //         } else {
    //             expect(ShippingInfoCard).toBeInTheDocument();
    //         }
    //     }, 300);
    // });

    // it('matches snapshot', () => {
    //     const run = false;
    //     if (run) {
    //         const tree = renderer.create(el).toJSON();
    //         expect(tree).toMatchSnapshot();
    //     }
    // });
});
