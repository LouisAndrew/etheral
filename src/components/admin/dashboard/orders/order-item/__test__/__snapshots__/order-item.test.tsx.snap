// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderItem Component matches snapshot 1`] = `
<div
<<<<<<< HEAD
  className="css-1nh1uvi"
=======
  className="css-1mt98ux"
>>>>>>> layout
>
  <div
    className="css-1b3c84b"
  >
    <div
      className="css-l9s608"
    >
      <h3
        className="css-ytgaz"
      >
        OrderId:
         
        <span
          style={
            Object {
              "color": "#553517",
            }
          }
        >
          asdadwda
        </span>
      </h3>
      <div
        className="css-1a2mf0e"
      >
        <svg
          dangerouslySetInnerHTML={
            Object {
              "__html": "<path d=\\"M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10s-4.477 10-10 10zm0-2a8 8 0 1 0 0-16a8 8 0 0 0 0 16zm0-9.414l2.828-2.829l1.415 1.415L13.414 12l2.829 2.828l-1.415 1.415L12 13.414l-2.828 2.829l-1.415-1.415L10.586 12L7.757 9.172l1.415-1.415L12 10.586z\\" fill=\\"currentColor\\"/>",
            }
          }
          focusable={false}
          height="1em"
          preserveAspectRatio="xMidYMid meet"
          style={
            Object {
              "transform": "rotate(360deg)",
              "verticalAlign": "-0.125em",
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        />
        Not paid
      </div>
    </div>
    <div
      className="css-1y35514"
      onClick={[Function]}
    >
      <svg
        dangerouslySetInnerHTML={
          Object {
            "__html": "<path d=\\"M12 13.172l4.95-4.95l1.414 1.414L12 16L5.636 9.636L7.05 8.222z\\" fill=\\"currentColor\\"/>",
          }
        }
        focusable={false}
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "transform": "rotate(360deg)",
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      />
    </div>
  </div>
</div>
`;
